tests/test_app.py

```diff
@@ -1,89 +1,87 @@
-    assert are_equal(maskvalues215, maskvalues215expected)
+            if b1 != b2:
+                return False
 
-    # Check feature extraction
-    # -----------------------------------------------------------------
-    features214 = tmpdir / "214/2012-08-01_terra_props.csv"
-    features214expected = expdir / "214/2012-08-01_terra_props.csv"
-    assert check_sums(features214, features214expected)
+            if not b1:  # End of file reached
+                return True
 
-    features215 = tmpdir / "215/2012-08-02_terra_props.csv"
-    features215expected = expdir / "215/2012-08-02_terra_props.csv"
-    assert check_sums(features215, features215expected)
 
-    # Check intermediate identification rounds
-    # -----------------------------------------------------------------
-    for doy in ["214", "215"]:
-        id_rounds = sorted(Path(tmpdir / doy).glob("*round*.tif"))
-        expected_rounds = sorted((expdir / doy).glob("*round*.tif"))
+def are_files_identical(file1, file2):
+    return filecmp.cmp(file1, file2, shallow=False)
 
-        for id_round, expected_round in zip(id_rounds, expected_rounds):
-            assert are_equal(id_round, expected_round)
 
+def are_images_identical(image1_path, image2_path):
+    with (
+        rasterio.open(str(image1_path)) as img1,
+        rasterio.open(str(image2_path)) as img2,
+    ):
+        return np.array_equal(img1.read(), img2.read())
 
-def getmaskvalues(path):
-    with open(path, "r") as f:
-        lines = f.readlines()
 
-    mask_values = [float(x) for x in lines[0].split()]
-    return mask_values
+def are_equal(p1, p2):
+    return Path(p1).read_bytes() == Path(p2).read_bytes()
 
 
-def group_files_by_extension(folder_path):
-    grouped_files = defaultdict(list)
-    folder = Path(folder_path)
+def check_sums(p1, p2):
+    s1 = pd.read_csv(p1).to_numpy().sum()
+    s2 = pd.read_csv(p2).to_numpy().sum()
+    return s1 == s2
 
-    for file in folder.iterdir():
-        if file.is_file():
-            extension = file.suffix
-            grouped_files[extension].append(file.name)
 
-    return dict(grouped_files)
+# Check final images
+def _test_output(tmpdir):
+    expdir = Path("tests/expected")
+    # Check all tif images, including final and identification rounds, and csv and txt files
+    # --------------------------------------------------------------------------------------
+    for d in getdirs(tmpdir):
+        for file in d.glob(
+            "*[!.png]"
+        ):  # Exclude png files as they are not created in the original code
+            expected = expdir / file.relative_to(tmpdir)
+            if file.suffix == ".csv":
+                assert check_sums(file, expected)
+                continue
+            if file.suffix == ".tif":
+                assert are_images_identical(file, expected)
+            assert files_are_equal(file, expected)
+            assert are_files_identical(file, expected)
+            assert are_equal(file, expected)
```

Comment:
It looks to me like these are each doing (almost?) the same thing. Is that right? Can you say something about why the `filecmp.cmp` function isn't sufficient? It would seem to me to be doing a byte-by-byte comparison which would make everything else redundant.

Reply:
@danielmwatkins [reported some differences](https://github.com/WilhelmusLab/ebseg/issues/27#issuecomment-2273865950) in the output files which I found a bit [surprising](https://github.com/WilhelmusLab/ebseg/issues/27#issuecomment-2284458499). I am trying different comparison methodologies here to boost our confidence in that the files are the same.

Reply:
Are the images you are comparing binary? Or are you comparing the labeled images?

Reply:
I am comparing every output file except for the histograms (`png`) as they are not available in the original repo. That is all `tif` images, the `mask_values.txt` file, and the props `csv` table.
```sh
$ tree temp/214
temp/214
├── 2012-08-01_ice_mask_bw.tif
├── 2012-08-01_identification_round_0.tif
├── 2012-08-01_identification_round_1.tif
├── 2012-08-01_identification_round_2.tif
├── 2012-08-01_identification_round_3.tif
├── 2012-08-01_identification_round_4.tif
├── 2012-08-01_identification_round_5.tif
├── 2012-08-01_terra_final.tif
├── 2012-08-01_terra_props.csv
├── cloud_mask_on_rgb.tif
├── ice_mask_hist.png # not this one
├── land_cloud_mask_on_rgb.tif
└── mask_values.txt
```

The only file that comes out with differences is the props table. Here is a sample distribution of the (very small) differences for the images with `doy = 214`:

```python
>>> (Pdb) nonzero_diffs_distribution.transpose()
                   False  True 
Unnamed: 0         345.0    0.0
label              345.0    0.0
area               345.0    0.0
centroid-0         345.0    0.0
centroid-1         345.0    0.0
axis_major_length  341.0    4.0
axis_minor_length  289.0   56.0
orientation        338.0    7.0
perimeter          305.0   40.0
intensity_mean     345.0    0.0
```

As can be seen from the table, only `axis_major/minor_length`, `orientation`, and `perimeter` are `anomalous_.